util/src/main/java/org/cryptomator/util/crypto/HubDeviceCryptor.java

```diff
@@ -0,0 +1,219 @@
+	private final KeyStore keyStore;
+
+	HubDeviceCryptor(KeyStore keyStore) {
+		try {
+			this.keyStore = keyStore;
+			this.keyStore.load(null);
+			if (!this.keyStore.containsAlias(DEFAULT_KEY_ALIAS)) {
+				var keyPairGenerator = KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_EC, DEFAULT_KEYSTORE_NAME);
+				var parameterSpec = new KeyGenParameterSpec //
+						.Builder(DEFAULT_KEY_ALIAS, KeyProperties.PURPOSE_DECRYPT | KeyProperties.PURPOSE_AGREE_KEY) //
+						.setAlgorithmParameterSpec(new ECGenParameterSpec("secp384r1")) //
+						.setDigests(KeyProperties.DIGEST_SHA256) //
+						.setUserAuthenticationRequired(false) //
+						.build();
+				keyPairGenerator.initialize(parameterSpec);
+				keyPairGenerator.generateKeyPair();
+			}
+		} catch (KeyStoreException | NoSuchAlgorithmException | NoSuchProviderException | InvalidAlgorithmParameterException | CertificateException | IOException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	public static HubDeviceCryptor getInstance() {
+		try {
+			var keyStore = KeyStore.getInstance(DEFAULT_KEYSTORE_NAME);
+			return new HubDeviceCryptor(keyStore);
+		} catch (KeyStoreException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	public static ECPrivateKey decryptUserKey(JWEObject jwe, String setupCode) throws InvalidJweKeyException {
+		try {
+			jwe.decrypt(new PasswordBasedDecrypter(setupCode));
+			return readKey(jwe, JWE_PAYLOAD_KEY_FIELD, HubDeviceCryptor::decodeECPrivateKey);
+		} catch (JOSEException e) {
+			throw new InvalidJweKeyException(e);
+		}
+	}
+
+	public static ECPrivateKey decryptUserKey(JWEObject jwe, PrivateKey deviceKey) {
+		try {
+			jwe.decrypt(new ECDHDecrypter(deviceKey, null, Curve.P_384));
+			return readKey(jwe, JWE_PAYLOAD_KEY_FIELD, HubDeviceCryptor::decodeECPrivateKey);
+		} catch (JOSEException e) {
+			throw new InvalidJweKeyException(e);
+		}
+	}
+
+	private static ECPrivateKey decodeECPrivateKey(byte[] encoded) throws KeyDecodeFailedException {
+		try {
+			var factory = KeyFactory.getInstance(EC_ALG);
+			var privateKey = factory.generatePrivate(new PKCS8EncodedKeySpec(encoded));
+			if (privateKey instanceof ECPrivateKey ecPrivateKey) {
+				return ecPrivateKey;
+			} else {
+				throw new IllegalStateException(EC_ALG + " key factory not generating ECPrivateKeys");
+			}
+		} catch (NoSuchAlgorithmException e) {
+			throw new IllegalStateException(EC_ALG + " not supported");
+		} catch (InvalidKeySpecException e) {
+			throw new KeyDecodeFailedException(e);
+		}
+	}
+
+	private static <T> T readKey(JWEObject jwe, String keyField, Function<byte[], T> rawKeyFactory) throws MasterkeyLoadingFailedException {
+		Preconditions.checkArgument(jwe.getState() == JWEObject.State.DECRYPTED);
+		var fields = jwe.getPayload().toJSONObject();
+		if (fields == null) {
+			Timber.tag("HubDeviceCryptor").e("Expected JWE payload to be JSON: " + jwe.getPayload());
+			throw new MasterkeyLoadingFailedException("Expected JWE payload to be JSON");
+		}
```

Comment:
I have used the same exceptions here as in the Deskop app and since we want to move the JWEHelper to the Cryptolib later, I want to be very close to the implication there.